File: LCOF/31-40/38/38.py
class Solution(object):
    def permutation(self, s):
        """
        :type s: str
        :rtype: List[str]
        """

        if not s:
            return [""]

        result = []
        visited = set()
        for i in range(len(s)):
            if s[i] not in visited:
                visited.add(s[i])
                rhs = self.permutation(s[:i]+ s[i+1:])
                for elem in rhs:
                    result.append(s[i]+elem)

        return result

File: LCOF/31-40/38/38.cpp
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    vector<string> permutation(string s) {
        if (s.empty())
            return vector<string>(1, "");

        vector<string> result;
        unordered_set<char> visited;

        for (int i = 0; i < s.size(); visited.insert(s[i++]))
            if (visited.find(s[i]) == visited.end())
                for (string &elem: permutation(s.substr(0, i)+s.substr(i+1, s.size()-i-1)))
                    result.push_back(s[i]+elem);

        return result;
    }
};